listings/applications/erc20/src/erc20_streaming.cairo

```diff
@@ -0,0 +1,118 @@
+#[starknet::contract]
+
+pub mod erc20_streaming {
+
+    // Import necessary modules and traits
+    use core::num::traits::Zero;
+    use starknet::get_caller_address;
+    use starknet::ContractAddress;
+    use starknet::LegacyMap;
+
+    #[storage]
+    struct Storage {
+        streams: LegacyMap<(ContractAddress, ContractAddress), Stream>,
+        erc20_token: ContractAddress,
+    }
+
+    #[derive(Copy, Drop, Debug, PartialEq)]
+    struct Stream {
+        start_time: u64,
+        end_time: u64,
+        total_amount: felt252,
+        released_amount: felt252,
+    }
+
+    #[event]
+    #[derive(Copy, Drop, Debug, PartialEq, starknet::Event)]
+    pub enum Event {
+        StreamCreated: StreamCreated,
+        TokensReleased: TokensReleased,
+    }
+
+    #[derive(Copy, Drop, Debug, PartialEq, starknet::Event)]
+    pub struct StreamCreated {
+        pub from: ContractAddress,
+        pub to: ContractAddress,
+        pub total_amount: felt252,
+        pub start_time: u64,
+        pub end_time: u64,
+    }
+
+    #[derive(Copy, Drop, Debug, PartialEq, starknet::Event)]
+    pub struct TokensReleased {
+        pub to: ContractAddress,
+        pub amount: felt252,
+    }
+
+    mod Errors {
+        pub const STREAM_AMOUNT_ZERO: felt252 = 'Stream amount cannot be zero';
+        pub const STREAM_ALREADY_EXISTS: felt252 = 'Stream already exists';
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, erc20_token: ContractAddress) {
+        self.erc20_token.write(erc20_token);
+    }
+
+    #[abi(embed_v0)]
+    impl IStreamImpl of super::IStream<ContractState> {
+        fn create_stream(
+            ref self: ContractState,
+            to: ContractAddress,
+            total_amount: felt252,
+            start_time: u64,
+            end_time: u64
+        ) {
+            assert(total_amount != felt252::zero(), Errors::STREAM_AMOUNT_ZERO);
+            let caller = get_caller_address();
+            let stream_key = (caller, to);
+            assert(self.streams.read(stream_key).start_time == 0, Errors::STREAM_ALREADY_EXISTS);
```

Comment:
This mean that one caller can only create an unique stream to a specific `to` address. You can replace this with a nonce system with a simple counter in storage to allow multiple streams with same caller and target.

Reply:
This was not resolved. You can add `from` inside `Stream` struct, change storage `stream` mapping key to uint, add `next_stream_id` uint as storage variable (like a counter) and consume/increment it for each new stream created.

Reply:
This is still not resolved.

Change storage to:
```cairo
    #[storage]
    struct Storage {
        streams: LegacyMap<u64, Stream>,
        next_stream_id: u64,  
    }
```

Then:
```cairo
let stream_key = self.next_stream_id;
self.next_stream_id.write(stream_key + 1);
```